drivers/stepper/ti/drv8424.c

```diff
@@ -0,0 +1,775 @@
+
+	data->max_velocity = velocity;
+	return 0;
+}
+
+static int drv8424_run(const struct device *dev, const enum stepper_direction direction,
+		       const uint32_t velocity)
+{
+	const struct drv8424_config *config = dev->config;
+	struct drv8424_data *data = dev->data;
+	int ret;
+
+	if (!data->enabled) {
+		return -ENODEV;
+	}
+
+	/* Set direction pin */
+	int dir_value = (direction == STEPPER_DIRECTION_POSITIVE) ? 1 : 0;
+
+	ret = gpio_pin_set_dt(&config->dir_pin, dir_value);
+	if (ret != 0) {
+		LOG_ERR("%s: Failed to set direction pin (error %d)", dev->name, ret);
+		return ret;
+	}
+
+	/* Lock interrupts while modifying settings used by ISR */
+	int key = irq_lock();
+
+	/* Set data used in counter interrupt */
+	data->constant_velocity = true;
+	data->constant_velocity_direction = direction;
+	data->is_moving = true;
+
+	/* Treat velocity 0 by not stepping at all */
+	if (velocity == 0) {
+		ret = counter_stop(config->counter);
+		if (ret != 0) {
+			LOG_ERR("%s: Failed to stop counter (error %d)", dev->name, ret);
+			data->is_moving = false;
+			goto end;
+		}
+		data->is_moving = false;
+		gpio_pin_set_dt(&config->step_pin, 0);
+		data->step_signal_high = false;
+	} else {
+		ret = drv8424_set_counter_frequency(dev, velocity * 2);
+		if (ret != 0) {
+			LOG_ERR("%s: Failed to set counter frequency (error %d)", dev->name, ret);
+			data->is_moving = false;
+			goto end;
+		}
+
+		/* Start counter */
+		ret = counter_start(config->counter);
+		if (ret != 0) {
+			LOG_ERR("%s: Failed to start counter (error %d)", dev->name, ret);
+			data->is_moving = false;
+			goto end;
+		}
+	}
+
+end:
+	irq_unlock(key);
+	return ret;
+}
+
+static int drv8424_set_micro_step_res(const struct device *dev,
+				      enum stepper_micro_step_resolution micro_step_res)
+{
+	const struct drv8424_config *config = dev->config;
+	struct drv8424_data *data = dev->data;
+	int ret;
+
+	int m0_value = 0;
+	int m1_value = 0;
+
+	/* 0: low
+	 * 1: high
+	 * 2: Hi-Z
+	 * 3: 330kΩ
+	 */
+	switch (micro_step_res) {
+	case STEPPER_MICRO_STEP_1:
+		m0_value = 0;
+		m1_value = 0;
+		break;
+	case STEPPER_MICRO_STEP_2:
+		m0_value = 2;
+		m1_value = 0;
+		break;
+	case STEPPER_MICRO_STEP_4:
+		m0_value = 0;
+		m1_value = 1;
+		break;
+	case STEPPER_MICRO_STEP_8:
+		m0_value = 1;
+		m1_value = 1;
+		break;
+	case STEPPER_MICRO_STEP_16:
+		m0_value = 2;
+		m1_value = 1;
+		break;
+	case STEPPER_MICRO_STEP_32:
+		m0_value = 0;
+		m1_value = 2;
+		break;
+	case STEPPER_MICRO_STEP_64:
+		m0_value = 2;
+		m1_value = 3;
+		break;
+	case STEPPER_MICRO_STEP_128:
+		m0_value = 2;
+		m1_value = 2;
+		break;
+	case STEPPER_MICRO_STEP_256:
+		m0_value = 1;
+		m1_value = 2;
+		break;
+	default:
+		return -EINVAL;
+	};
+
+	/* Reset m0 pin as it may have been disconnected. */
+	ret = gpio_pin_configure_dt(&config->m0_pin, GPIO_OUTPUT_INACTIVE);
+	if (ret != 0) {
+		LOG_ERR("%s: Failed to reset m0_pin (error: %d)", dev->name, ret);
+		drv8424_microstep_recovery(dev);
```

Comment:
It's best if this fallback mechanism is exposed to user via include/zephyr/drivers/stepper/stepper_drv8424.h, If the setting of the microstep fails then return -EIO and then the user can decide to use this fallback mechanism